File: ch1/modoki01/java/ServerThread.java
import java.io.*;
import java.net.*;
import java.util.*;
import java.text.*;

public class ServerThread implements Runnable {
    private static final String DOCUMENT_ROOT = "./../";
    private Socket socket;

    // InputStreamからのバイト列を、行単位で読み込むユーティティメソッド
    private static String readLine(InputStream input) throws Exception {
        int ch;
        String ret = "";
        while ((ch = input.read()) != -1) {
            if (ch == '\r') {
                // 何もしない
            } else if (ch == '\n') {
                break;
            } else {
                ret += (char)ch;
            }
        }
        if (ch == -1) {
            return null;
        } else {
            return ret;
        }
    }

    // 1行の文字列を、バイト列としてOutputStreamに書き込む
    // ユーティリティメソッド
    private static void writeLine(OutputStream output, String str)
        throws  Exception {
        for (char ch : str.toCharArray()) {
            output.write((int)ch);
        }
        output.write((int)'\r');
        output.write((int)'\n');
    }

    // 現在時刻から、HTTP標準に合わせてフォーマットされた日付文字列を返す
    private static String getDateStringUtc() {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        DateFormat df = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss",
                                             Locale.US);
        df.setTimeZone(cal.getTimeZone());
        return df.format(cal.getTime()) + " GMT";
    }

    // 拡張子とContent-Typeの対応表
    private static final HashMap<String, String> contentTypeMap =
        new HashMap<String, String>() {{
            put("html", "text/html");
            put("htm", "text/html");
            put("txt", "text/plain");
            put("css", "text/css");
            put("png", "image/png");
            put("jpg", "image/jpeg");
            put("jpeg", "image/jpeg");
            put("gif", "image/gif");
        }
    };
    // 拡張子を受け取りContent-Typeを返す
    private static String getContentType(String ext) {
        String ret = contentTypeMap.get(ext.toLowerCase());
        if (ret == null) {
            return "application/octet-stream";
        } else {
            return ret;
        }
    }

    @Override
    public void run() {
        OutputStream output;
        try {
            InputStream input = socket.getInputStream();

            String line;
            String path = null;
            String ext = null;
            while ((line = readLine(input)) != null) {
                if (line.equals(""))
                    break;
                if (line.startsWith("GET")) {
                    path = line.split(" ")[1];
                    String[] tmp = path.split("¥¥.");
                    ext = tmp[tmp.length - 1];
                }
            }
            output = socket.getOutputStream();
            // レスポンスヘッダを返す
            writeLine(output, "HTTP/1.1 200 OK");
            writeLine(output, "Date: " + getDateStringUtc());
            writeLine(output, "Server: Modoki/0.1");
            writeLine(output, "Connection: close");
            writeLine(output, "Content-Type: " + getContentType(ext));
            writeLine(output, "");

            // レスポンスボディを返す
            try (FileInputStream fis
                 = new FileInputStream(DOCUMENT_ROOT + path);) {
                int ch;
                while ((ch = fis.read()) != -1) {
                    output.write(ch);
                }
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        } finally {
            try {
                socket.close();
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    ServerThread(Socket socket) {
        this.socket = socket;
    }
}

File: ch1/modoki01/modoki01/server_thread.py
import datetime
import threading
import os.path


class ServerThread(threading.Thread):

    def __init__(self, document_root, sock):
        super(ServerThread, self).__init__()

        self.__DOCUMENT_ROOT = document_root
        self.sock = sock

    @property
    def document_root(self):
        return self.__DOCUMENT_ROOT

    def run(self):

        path = None
        ext = None

        with self.sock.makefile() as input_stream:
            line = read_line(input_stream)

            while not line.isspace():

                if (line == ""):
                    break

                if (line.startswith("GET")):
                    path = line.split(" ")[1]
                    root, ext_with_dot = os.path.splitext(path)
                    ext = ext_with_dot.lstrip(".")

                line = read_line(input_stream)

            write_line(self.sock, "HTTP/1.1 200 OK")
            write_line(self.sock, "Date: " + get_date_string_utc())
            write_line(self.sock, "Server: Modoki/0.1")
            write_line(self.sock, "sockection: close")
            write_line(self.sock, "Content-type: " + get_content_type(ext))
            write_line(self.sock, "")

            with open(self.__DOCUMENT_ROOT + path, "rb") as fp:
                self.sock.send(fp.read())


def write_line(stream, line):
    stream.send((line + "\r\n").encode())


def read_line(stream):
    return stream.readline()


def get_date_string_utc():
    now = datetime.datetime.utcnow()
    return "{0:%a, %d %b %Y %H:%M:%S GMT}".format(now)


def get_content_type_map():

    type_map = {}

    type_map["html"] = "text/html"
    type_map["htm"] = "text/html"
    type_map["txt"] = "text/plain"
    type_map["css"] = "text/css"
    type_map["pnt"] = "image/png"
    type_map["jpg"] = "image/jpeg"
    type_map["jpeg"] = "image/jpeg"
    type_map["gif"] = "image/gif"

    return type_map


def get_content_type(ext):

    ret = get_content_type_map()[ext.lower()]

    if ret is None:
        return "application/octet-stream"
    else:
        return ret
